### DISCUSSION TITLE
Lesson 10: testing rationale

### DISCUSSION THREAD

**Discussion by leopashov** [smartcontractkit/full-blockchain-solidity-course-py#856] [Q&A] [answered]
Hi,

Thanks again for the course, it's awesome!

I just wanted to ask about the reasoning behind the line:
`assert tx is not True'
in the approve_erc20 unit test, which tests the approve function in lesson 10.
I can see the tx type is <class 'brownie.network.transaction.TransactionReceipt'>
and the tx result is a hash, but how would we know to write 'tx is not True' (which seems strange on its own, as asserting 'tx is False' gives an error)?

Thank you! :)

**Comment by cromewar** [comment, marked as answer]
Hello @leopashov this is due to the the response withing the transaction, if it success the transaction goes True, if is reverted by any malfunction the tx goes false.

**Reply by leopashov** [reply]
But then don't we want the transaction to show True? As we want to assert a successful tx?

**Reply by cromewar** [reply]
Is because you are not getting the whole transaction just the state of it, TransactionReceip if the reception passed or no. 
You can see more info on the transactions using this link: https://eth-brownie.readthedocs.io/en/stable/core-transactions.html

**Reply by leopashov** [reply]
Ahh I see, thank you very much for this info! :)

**Comment by Niferu** [comment]
Hello @cromewar and @leopashov ,

I had the same question and i found this topic, but it is still unclear to me as I'm pretty new to programming at all and to blockchain.

1. First of all could you please clarify why "is not True" is not the same as "is False" and test throws error for "is False"?
2. If I got it correct we are checking If "tx" throws any malfunction and if it does, test should pass. Below screen shows "tx.info()" function result and in "Event In This Transaction" we can see "(unknown)" data, so is it the malfunction? If it is why do we even search for malfunction to confirm erc20 approval function passed, it doesn't make sense to me.
![image](https://user-images.githubusercontent.com/110537615/187568465-bf2f2d12-0d84-4eea-88df-823b2efaa366.png)

**Reply by cromewar** [reply]
Hello @Neftyr 
1. Well this is because the Tx responses are not synchronous, so using Tx is False will be incorrect as it will throw an error, when the test passes the Tx is neither True or False, we don't know that because we are not getting the whole transaction, just the transaction receipt.

In regular programming logic False and not True are the same, but here this works because as I said we don't want the whole transact just the receipt, and as it is an asynchronous process it'll take time.

2. Well this is also related to the last question, we don't need to check out the whole transaction in order to see what happened, because as you can see the data is encrypted, but the internal EVM know the transaction has something which is not supposed to be there, so it immediately throws an error, this error make the tx receipt to return the error and in consequence we know the test is working as it should.

I hope this might help to clarify the problem.

**Reply by Niferu** [reply]
Hello @cromewar 

Thanks again for your help! Point 1 is crystal clear to me now (I have checked and stating "is not False" works also and test passes), thanks again for great explanation with regular programming comparison. Regarding point 2 let me explain it to you in way i see it and please confirm if i understood it correct.

1. We send transaction (Tx)
2. As it is approval and not standard transaction data in it is encrypted
3. Even if it is encrypted EVM catch that there is something wrong with this Tx as it is not standard transaction and throws error
4. Tx receipt return that error to us (btw where can we see this error message? Or what does it say?)
5. As Tx receipt took action and passed error we know that tx receipt has been created and exists, which confirms our approval passed as we wanted to receive tx receipt.

Is my above understanding correct? Sorry for hollowing but I'm trying to understand things in this course and things correlated to it and not just follow Patrick brainless.

**Reply by cromewar** [reply]
Yes you are right you got it, but actually I have to confess my explanation was kinda incomplete as I skipped some parts of it, the technical ones as you must have a deeper understanding on python and brownie, however if you are interested on learn the technicalities, please don't forget to check out the brownie docs to expand your knowledge: https://eth-brownie.readthedocs.io/en/stable/core-transactions.html

> btw where can we see this error message? Or what does it say?

Yes, you can get the  `tx.revert_msg` to know what happened and then, based on the message decide what you want to do.

**Reply by Niferu** [reply]
Thanks a lot Sir @cromewar you are amazing!
